crates/codegen/syntax_templates/src/shared/cursor.rs

```diff
@@ -0,0 +1,342 @@
+use std::rc::Rc;
+
+use super::{
+    cst::{Node, RuleNode, TokenNode},
+    kinds::{RuleKind, TokenKind},
+    text_index::{TextIndex, TextRange},
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorPathElement {
+    rule_node: Rc<RuleNode>,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CursorLeaf {
+    node: Node,
+    child_number: usize,
+    text_offset: TextIndex,
+}
+
+impl CursorLeaf {
+    pub fn text_range(&self) -> TextRange {
+        let start = self.text_offset;
+        let end = start + self.node.text_len();
+        start..end
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Cursor {
+    path: Vec<CursorPathElement>,
+    leaf: Option<CursorLeaf>,
+}
+
+impl Cursor {
+    pub(crate) fn new(node: Node) -> Self {
+        Self {
+            path: vec![],
+            leaf: Some(CursorLeaf {
+                node,
+                child_number: 0,
+                text_offset: Default::default(),
+            }),
+        }
+    }
+
+    // Unlike clone, this re-roots at the current node.
+    // It does preserve the correct text offset however,
+    // even though the path is reset.
+    pub fn spawn(&self) -> Self {
```

Comment:
thinking out loud, is `spawn` common terminology for cursors? vs, for example, `fork`?

Reply:
It is for me. `fork()` is overloaded I think.